### DISCUSSION TITLE
Do all features work without a PAT?

### DISCUSSION THREAD

**Discussion by lukasbestle** [upptime/upptime#231] [Q&A]
I'm planning to use Upptime without a GitHub personal access token.

I read that Upptime uses the always available `GITHUB_TOKEN` if no PAT is configured. Does this mode support all features or are there limitations that can only be circumvented with a PAT?

**Comment by AnandChowdhary** [comment]
At this point, it's unlikely that Upptime will work for you without a PAT. We do things like pushing to GitHub Pages and using the GitHub API, which is not possible using the `GITHUB_TOKEN`.

Could you share why you're thinking about not using a PAT? Perhaps we can come up with a creative solution!

**Reply by lukasbestle** [reply]
Thanks for your reply!

My plan is to use Upptime for a GitHub organization, so a PAT from a personal GitHub account of a team member wouldn't be an option (access wouldn't be limited to the organization and the workflows would break if the team member leaves the organization).
The only option would be a bot user. We already have one for another use-case, but the bot user has write access to several other repos.

For security, I want to limit the access of the PAT to just the Upptime use-case, but as far as I can tell, PAT always carry the full "personal access" of the user and can't be limited to specific repos.

Are you sure the GitHub API can't be used with the `GITHUB_TOKEN`? As far as I read, both pushing to the current repo and accessing the GitHub API is possible with the token.

**Reply by AnandChowdhary** [reply]
> Are you sure the GitHub API can't be used with the GITHUB_TOKEN?

No, it was an assumption that is likely incorrect.

If you don't want the status website and only want to use the GitHub issues and README part, you can probably do that with the default `GITHUB_TOKEN`. However, publishing to GitHub Pages isn't supported with that (yet) but this could change soon: https://github.com/maxheld83/ghpages/issues/1

**Reply by lukasbestle** [reply]
Ah, I see!

A user mentions that GitHub Pages will be deployed when pushing using an SSH deployment key: https://github.com/maxheld83/ghpages/issues/1#issuecomment-459577685
Have you tested that so far? Could it be an alternative?

**Reply by AnandChowdhary** [reply]
Yes, I now think that only updating the template needs the `GH_PAT`, everything else can be done using the default `GITHUB_TOKEN`.

**Reply by asaf** [reply]
@AnandChowdhary  The [JamesIves/github-pages-deploy-action](https://github.com/JamesIves/github-pages-deploy-action) github action deploys to `gh_pages` using the provided repository scoped GitHub token. why wouldn't it work in upptime scenario ?

**Reply by se-bastiaan** [reply]
Just to put this somewhere:

All functionality works perfectly fine when using the GitHub token (see [tests](https://github.com/se-bastiaan/statuspage-test/actions)). The only real reason the PAT is necessary is because of the [template update action](https://github.com/upptime/uptime-monitor/blob/master/src/workflows/update-template.yml) that is executed periodically.

This does depend on one setting though, which determines the default [Github Action permissions](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token) which is set to permissive by default. So if you have not changed that setting the token should work out of the box.

The reason the Github token does not work in the update template actions is because it is missing the workflow [scope](https://docs.github.com/en/developers/apps/building-oauth-apps/scopes-for-oauth-apps), because the actions are designed to not be able to edit themselves.

**Reply by olivierdalang** [reply]
Thank @se-bastiaan !

Indeed much better to use GITHUB_TOKEN instead of a PAT (one must really not value his github account too much to dare using a PAT with so many github actions...). IMO that should be the default for upptime, if some features are not available with GITHUB_TOKEN then they should not be included.

For anyone else interested, here are the exact changes needed : https://github.com/se-bastiaan/statuspage-test/commit/56d7ec8eecf1e84e71dd1b972345a72291193c23

**Reply by Andre601** [reply]
> Just to put this somewhere:
> 
> All functionality works perfectly fine when using the GitHub token (see [tests](https://github.com/se-bastiaan/statuspage-test/actions)). The only real reason the PAT is necessary is because of the [template update action](https://github.com/upptime/uptime-monitor/blob/master/src/workflows/update-template.yml) that is executed periodically.
> 
> This does depend on one setting though, which determines the default [Github Action permissions](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token) which is set to permissive by default. So if you have not changed that setting the token should work out of the box.
> 
> The reason the Github token does not work in the update template actions is because it is missing the workflow [scope](https://docs.github.com/en/developers/apps/building-oauth-apps/scopes-for-oauth-apps), because the actions are designed to not be able to edit themselves.

Want to point out that the template update action is gone (Or at least changed location, as the link leads to a 404).

Also, like mentioned in my reply here, are normal github tokens able to trigger a workflow/repository dispatch event, so if what you're saying is correct should the default token now be capable to run completely alone without any separate account and PAT needed.

**Reply by se-bastiaan** [reply]
It was just moved. https://github.com/upptime/upptime/blob/master/.github/workflows/update-template.yml

**Comment by olivierdalang** [comment]
@AnandChowdhary Would you consider changing uptime to use GITHUB token instead of a PAT ? In my experience, everything works fine except the "update template" workflow (see below). That workflow could be replace by a manual update procedure (which most ppl would probably prefer over an automatic update for a production monitoring tool).

Failure on update-template.yml (the workflow still shows as successful):
```
! [remote rejected] master -> master (refusing to allow a GitHub App to create or update workflow `.github/workflows/graphs.yml` without `workflows` permission)
```

**Reply by QuLogic** [reply]
I'd definitely rather it use a GitHub token and then file an issue when the template needed an update.

**Comment by Andre601** [comment]
So, just to have the up-to-date info:

What actions, features, etc. require a PAT and cannot work with the default token generated for GitHub Actions?

From personal experience can I already tell that pushing to the `gh-pages` branch on the same repo works fine with the default token (Only requires a PAT for cross-repo pushing) and pretty much everything else I know of (creating/managing issues, pushing updates to files, etc) works fine too.

So, what there is left?

**Reply by se-bastiaan** [reply]
Updating the GitHub Actions. See https://github.com/upptime/upptime/discussions/231#discussioncomment-1562717

**Reply by Andre601** [reply]
Okay. And how exactly does this affect the entire thing now?
Like your reply there is very detailed and all, but in what way would not using that template update thing impact the entire thing negatively?
Like would I get errors about failed runs? Would it simply not run at all? Should I use that action anyways?

It doesn't matter this much to me as I use a dedicated account for it anyway, but would still like to know if there would be any major complications (errors, missing page updates, limitations, etc) for switching to a GH-token only solution.

**Comment by Andre601** [comment]
I would like to take the time and quickly mention, that the default `GITHUB_TOKEN` is now able to trigger other workflows through a `workflow_dispatch`/`repository_dispatch` event trigger, meaning that upptime may now be able to completely work with just the default token used and no extra account required.

Of course, this needs to be tested and from what I understand would I just need to remove the PAT from the repository secrets?

**Reply by se-bastiaan** [reply]
It may be able to trigger other workflows. But the update template action still won't work. Since actions aren't allowed to update the workflows and the scope is still missing.

```
Removed legacy .github/workflows
Added new .github/workflows
Removed template files
Removed old data from api
Removed old data from graphs
Removed old data from history
[master fd0f67a] :arrow_up: Update @upptime to v1.27.0
 9 files changed, 22 insertions(+), 25 deletions(-)
 delete mode 100644 .templaterc.json
To https://github.com/se-bastiaan/statuspage-test
 ! [remote rejected] master -> master (refusing to allow a GitHub App to create or update workflow `.github/workflows/graphs.yml` without `workflows` permission)
error: failed to push some refs to 'https://github.com/se-bastiaan/statuspage-test'
All done!
```

**Reply by Andre601** [reply]
I can see a few possible solutions here to work around this:

- Just don't update the actions. Let the user do it.
- Add a system that informs the maintainer(s) through opening an issue about the new versions, so that they can update it.
- Make the actions actual, separate GitHub Actions, so that a bot like Dependabot or Renovate can update them automatically.
- From what I can see is @upptime-bot used to update workflows. Maybe create an App under the upptime org and allow people to invite it to the repositories? With the right permissions could it still update the workflows. And actions can use a bot-account thanks to actions such as tibdex/github-app-token, so there would certainly be a solution

In either case is an account really not needed anymore I would say... There are solutions to get around this, but they require updates to the overall upptime structure (Which I myself can't give sadly).

**Reply by se-bastiaan** [reply]
If a different solution is implemented the requirement of using the PAT is removed. The question here is 'Do all features work without a PAT?' and the answer to that is still _no_ regardless of what the workaround is someone may want to use.

The only feasible solution is opening a PR with the changes instead of trying to push them directly to the repository.
